### PR TITLE
Fix bug with if nested, false condition is evaluated if it is an if function

### PR DESCRIPTION
**Pull request by laurentyhuel**
Fix bug with if nested, false condition is evaluated if it is an if function , related to issue : #210

This PR  use a different way to fix this bug  that PR #211
 as asked in comment https://github.com/uklimaschewski/EvalEx/pull/211#pullrequestreview-273455041 and https://github.com/uklimaschewski/EvalEx/pull/211#issuecomment-520212442

### CODE REVIEW

File: src/main/java/com/udojava/evalex/LazyIfNumber.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package com.udojava.evalex;
```
Comment (RobertZenz) [inline src/main/java/com/udojava/evalex/LazyIfNumber.java]: The license header is missing.

File: src/main/java/com/udojava/evalex/LazyIfNumber.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.udojava.evalex.Expression.LazyNumber;
+import java.math.BigDecimal;
+import java.util.List;
+
+public class LazyIfNumber implements LazyNumber {
+
+  List<LazyNumber> lazyParams;
```
Comment (RobertZenz) [inline src/main/java/com/udojava/evalex/LazyIfNumber.java]: I'd make this `protected` to give subclasses a chance to access it.

File: src/main/java/com/udojava/evalex/LazyIfNumber.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  public BigDecimal eval() {
+    BigDecimal result = lazyParams.get(0).eval();
+    assertNotNull(result);
+    boolean isTrue = result.compareTo(BigDecimal.ZERO) != 0;
+    return isTrue ? lazyParams.get(1).eval() : lazyParams.get(2).eval();
+  }
+
+  @Override
+  public String getString() {
+    return "IF";
```
Comment (RobertZenz) [inline src/main/java/com/udojava/evalex/LazyIfNumber.java]: I believe that other implementations are either returning an actual `String` representation of the operation, or the result as `String`. Could you double check that?

File: src/test/java/com/udojava/evalex/TestLazyIf.java [hunk 4]
```diff
@@ -26,7 +26,23 @@ public void testLazyIfWithNestedFunction() {
...
+
+  @Test
+  public void testLazyIfWithNested() {
+    Expression expression = new Expression("if(a=0,0,if(5/a>3,2,4))");
+    expression.setVariable("a", new BigDecimal(0));
+    Assert.assertEquals(expression.eval(),new BigDecimal(0));
+  }
+
+  @Test(expected = ArithmeticException.class)
+  public void testLazyIfWithNested2() {
```
Comment (RobertZenz) [inline src/test/java/com/udojava/evalex/TestLazyIf.java]: You should name it something like `testLazyIfWithNestedFailingIf` to make it clear that this test is supposed to be the test of a failure case.

Comment (laurentyhuel) [reply src/test/java/com/udojava/evalex/TestLazyIf.java]: Hello @RobertZenz, thanks for your review. I made changes

File: src/main/java/com/udojava/evalex/LazyIfNumber.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.udojava.evalex.Expression.LazyNumber;
+import java.math.BigDecimal;
+import java.util.List;
+
+public class LazyIfNumber implements LazyNumber {
+
+  private List<LazyNumber> lazyParams;
+
+  LazyIfNumber(List<LazyNumber> lazyParams) {
```
Comment (RobertZenz) [inline src/main/java/com/udojava/evalex/LazyIfNumber.java]: Sorry, I missed this completely on my first review. Having a public class but a package-private constructor is slightly problematic from an API point of view. As it allows everyone to see the class, but not to use it.

I'd suggest to make it `public` or the whole class `package-private`...but I'd need to do some more thinking about that one, so I'd go with `public`.

**Comment by uklimaschewski** [comment]
I hope to find some time this weekend to merge this PR and have a look at the other issues.
I have just started working at a new job, so I am quite busy at the moment. 🙂

**Comment by RobertZenz** [comment]
I just verified that these changes are passing all tests (including those in my project).

**Comment by laurentyhuel** [comment]
@RobertZenz @uklimaschewski Could you create a new version 2.2 with this fix ?

**Comment by uklimaschewski** [comment]
@laurentyhuel Version 2.2 was built and should pop up at Maven Central in the next while.

**Pull request merged**